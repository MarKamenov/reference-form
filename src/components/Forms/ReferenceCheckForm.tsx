import { ChangeEvent, FormEvent, Fragment, useRef, useState } from "react";
import { EmployerKeys, FormState } from "src/types";
import { FormField } from './FormField'
import referenceCheckFormStyles from './ReferenceCheckForm.module.scss'
import { saveReferenceForm } from "src/services/formsService";

export const ReferenceCheckForm = () => {
    const [formData, setFormData] = useState<FormState>({
        personal: {
            firstName: '',
            lastName: '',
            currentAddress: '',
        },
        employers: [{ name: '', startDate: '', endDate: '' }],
    });
    const [saveError, setSaveError] = useState<{ message: string; stack: string } | null>(null);
    const savedValuesRef = useRef<FormState>(formData);

    const handlePersonalChange = (e: ChangeEvent<HTMLInputElement>) => {
        const { name, value } = e.target;
        setFormData(prev => ({
            ...prev,
            personal: { ...prev.personal, [name]: value },
        }));
    };

    const handleEmployerChange = (index: number, field: EmployerKeys, value: string) => {
        setFormData(prev => ({
            ...prev,
            employers: prev.employers.map((emp, i) =>
                i === index ? { ...emp, [field]: value } : emp
            ),
        }));
    };

    const revertFormData = () => {
        setFormData(savedValuesRef.current);
    }

    const handleSubmit = async (e: FormEvent) => {
        e.preventDefault();
        savedValuesRef.current = formData;
        try {
            const data = await saveReferenceForm(formData)
            console.log('Data saved', data)
            setSaveError(null)
        } catch (error: unknown) {
            console.error('Error:', error);
            setSaveError(error as { message: string; stack: string });

        }
    };

    return (
        <div>
            {saveError && <p>{saveError.message}</p>}
            <form aria-label="signup-form" data-testid="form" onSubmit={handleSubmit}>
                <fieldset className={referenceCheckFormStyles.fieldset}>
                    <legend>Personal</legend>
                    <div className={referenceCheckFormStyles.field}>
                        <FormField
                            label="First Name"
                            type='text'
                            id="firstName"
                            name="firstName"
                            value={formData.personal.firstName}
                            handleChange={handlePersonalChange}
                            required />
                    </div>
                    <div className={referenceCheckFormStyles.field}>
                        <FormField
                            label="Last Name"
                            type='text'
                            id="lastName"
                            name="lastName"
                            value={formData.personal.lastName}
                            handleChange={handlePersonalChange}
                            required />
                    </div>
                    <div className={referenceCheckFormStyles.field}>
                        <FormField
                            label="Address"
                            type='text'
                            id="currentAddress"
                            name="currentAddress"
                            value={formData.personal.currentAddress}
                            handleChange={handlePersonalChange}
                            required />
                    </div>
                </fieldset>

                <fieldset className={referenceCheckFormStyles.fieldset}>
                    <legend>Employer</legend>
                    {formData.employers.map((employer, index) => (
                        <Fragment key={`${index}${employer.endDate}`} >
                            <div className={referenceCheckFormStyles.field}>
                                <FormField
                                    label="Employer Name"
                                    type='text'
                                    id={`employerName-${index}`}
                                    name="name"
                                    value={employer.name}
                                    handleChange={(e) => handleEmployerChange(index, e.target.name as EmployerKeys, e.target.value)}
                                    required />
                            </div>
                            <div className={`${referenceCheckFormStyles.date__wrapper} ${referenceCheckFormStyles.field}`}>
                                <div className={referenceCheckFormStyles.date__wrapper__form}>
                                    <FormField
                                        label="Start Date"
                                        type='date'
                                        id={`startDate-${index}`}
                                        name="startDate"
                                        value={employer.startDate}
                                        handleChange={(e) => handleEmployerChange(index, e.target.name as EmployerKeys, e.target.value)}
                                        required />
                                </div>
                                <div className={referenceCheckFormStyles.date__wrapper__form}>
                                    <FormField
                                        label="End Date"
                                        type='date'
                                        id={`endDate-${index}`}
                                        name="endDate"
                                        value={employer.endDate}
                                        handleChange={(e) => handleEmployerChange(index, e.target.name as EmployerKeys, e.target.value)}
                                    />
                                </div>
                            </div>
                        </Fragment>
                    ))}
                </fieldset>
                <div className={referenceCheckFormStyles.actions}>
                    <button type='button' onClick={revertFormData} className={referenceCheckFormStyles.actions__cancel}>
                        Cancel
                    </button>
                    <button className={referenceCheckFormStyles.actions__submit} type="submit">
                        Submit
                    </button>
                </div>
            </form>
        </div>
    );
};

