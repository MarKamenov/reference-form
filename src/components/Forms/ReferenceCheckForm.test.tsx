import { render } from '@testing-library/react';
import { screen } from '@testing-library/dom'

import { ReferenceCheckForm } from './ReferenceCheckForm';

describe('FormField component', () => {

    test('should render component', () => {
        const { container } = render(<ReferenceCheckForm />)

        expect(container).toBeInTheDocument();
    });

    test('should match snapshot', () => {
        const { container } = render(<ReferenceCheckForm />);

        expect(container).toMatchSnapshot();
    });

    test('has correct input value', () => {
        render(<ReferenceCheckForm />)

        expect(screen.getByTestId('form')).toHaveFormValues({
            firstName: '',
            lastName: '',
            currentAddress: '',
            name: '',
            startDate: '',
            endDate: ''

        })
    })
});

