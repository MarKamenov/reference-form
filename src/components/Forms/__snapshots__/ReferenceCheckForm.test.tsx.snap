// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField component should match snapshot 1`] = `
<div>
  <div>
    <form
      aria-label="signup-form"
      data-testid="form"
    >
      <fieldset
        class="fieldset"
      >
        <legend>
          Personal
        </legend>
        <div
          class="field"
        >
          <label
            for="firstName"
          >
            First Name
          </label>
          <input
            data-testid="input"
            id="firstName"
            name="firstName"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="field"
        >
          <label
            for="lastName"
          >
            Last Name
          </label>
          <input
            data-testid="input"
            id="lastName"
            name="lastName"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="field"
        >
          <label
            for="currentAddress"
          >
            Address
          </label>
          <input
            data-testid="input"
            id="currentAddress"
            name="currentAddress"
            required=""
            type="text"
            value=""
          />
        </div>
      </fieldset>
      <fieldset
        class="fieldset"
      >
        <legend>
          Employer
        </legend>
        <div
          class="field"
        >
          <label
            for="employerName-0"
          >
            Employer Name
          </label>
          <input
            data-testid="input"
            id="employerName-0"
            name="name"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="date__wrapper field"
        >
          <div
            class="date__wrapper__form"
          >
            <label
              for="startDate-0"
            >
              Start Date
            </label>
            <input
              data-testid="input"
              id="startDate-0"
              name="startDate"
              required=""
              type="date"
              value=""
            />
          </div>
          <div
            class="date__wrapper__form"
          >
            <label
              for="endDate-0"
            >
              End Date
            </label>
            <input
              data-testid="input"
              id="endDate-0"
              name="endDate"
              type="date"
              value=""
            />
          </div>
        </div>
      </fieldset>
      <div
        class="actions"
      >
        <button
          class="actions__cancel"
          type="button"
        >
          Cancel
        </button>
        <button
          class="actions__submit"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;
